envision/server.py

```diff
@@ -54,30 +64,214 @@ def options(self):
+        self._data = data
+        self._size = sys.getsizeof(data)
+        self.next_ = next_
+
+    @property
+    def timestamp(self):
+        return self._timestamp
+
+    @property
+    def data(self):
+        return self._data
+
+    @data.setter
+    def data(self, data: str):
+        self._data = data
+        self._size = sys.getsizeof(data)
+
+    @property
+    def size(self):
+        return self._size
+
+
+class Frames:
+    def __init__(self, max_capacity_mb=500):
+        self._origin_timestamp = None
+        self._max_capacity = max_capacity_mb
+
+        # XXX: Index of timestamp in `self._timestamps` matches the index of the
+        # frame with the same timestamp in `self._frames`.
+        self._timestamps = []
+        self._frames = []
+
+    @property
+    def start_frame(self):
+        if len(self._frames) == 0:
+            return None
+
+        return self._frames[0]
+
+    @property
+    def start_time(self):
+        return self._origin_timestamp
+
+    @property
+    def elapsed_time(self):
+        return self._frames[-1].timestamp - self._origin_timestamp
+
+    def append(self, frame: Frame):
+        if len(self._frames) == 0:
+            self._origin_timestamp = frame.timestamp
+
+        self._enforce_max_capacity()
+
+        if len(self._frames) >= 1:
+            self._frames[-1].next_ = frame
+
+        self._frames.append(frame)
+        self._timestamps.append(frame.timestamp)
+
+    def __call__(self, timestamp):
+        """Finds the nearest frame according to the given timestamp."""
+        frame_idx = bisect.bisect_left(self._timestamps, timestamp)
+        if frame_idx >= len(self._frames):
+            frame_idx = -1
+
+        return self._frames[frame_idx]
+
+    def _enforce_max_capacity(self):
+        """Sample random frames and clear their data to ensure we're under the max
+        capacity size.
+        """
+        sizes = [frame.size for frame in self._frames]
+        while sum(sizes) * 1e-6 > self._max_capacity:
+            # XXX: randint(1, ...), we skip the start frame because that is a "safe"
+            #      frame we can always rely on being available.
+            idx_to_delete = random.randint(1, len(self._frames) - 1)
+            sizes[idx_to_delete] = 0
+            self._frames[idx_to_delete].data = None
+
+
+class WebClientRunLoop:
+    """The run loop is like a "video player" for the simulation. It supports seeking
+    and playback. The run loop wraps the web client handler and pushes the frame
+    messages to it as needed.
+    """
+
+    def __init__(self, frames, web_client_handler, timestep_sec, seek=None):
+        self._frames = frames
+        self._client = web_client_handler
+        self._timestep_sec = timestep_sec
+        self._seek = seek
+        self._thread = None
+
+    def seek(self, offset_seconds):
+        self._seek = offset_seconds
+
+    def stop(self):
+        if self._thread:
+            self._thread.join(timeout=3)
```

Comment:
I cannot tell where this is called. Might want to do something if this fails so the thread does not hang around at the end.